dkg/pedersen/controller/action.go

```diff
@@ -230,57 +230,131 @@ func (a decryptAction) Execute(ctx node.Context) error {
+
+	err := ctx.Injector.Resolve(&actor)
 	if err != nil {
-		return "", xerrors.Errorf("failed to marshal k: %v", err)
+		return xerrors.Errorf(resolveActorFailed, err)
 	}
 
-	cbuff, err := c.MarshalBinary()
+	encrypted := ctx.Flags.String("encrypted")
+
+	k, _, err := decodeEncrypted(encrypted)
 	if err != nil {
-		return "", xerrors.Errorf("failed to marshal c: %v", err)
+		return xerrors.Errorf("failed to decode encrypted str: %v", err)
 	}
 
-	encoded := hex.EncodeToString(kbuff) + separator +
-		hex.EncodeToString(cbuff) + separator +
-		hex.EncodeToString(remainder)
+	publicKey := ctx.Flags.String("pubk")
 
-	return encoded, nil
-}
+	pk, err := decodePublicKey(publicKey)
+	if err != nil {
+		return xerrors.Errorf("failed to decode public key str: %v", err)
+	}
 
-func decodeEncrypted(str string) (k kyber.Point, c kyber.Point, err error) {
-	parts := strings.Split(str, separator)
-	if len(parts) < 2 {
-		return nil, nil, xerrors.Errorf("malformed encoded: %s", str)
+	xhatenc, err := actor.Reencrypt(k, pk)
+	if err != nil {
+		return xerrors.Errorf("failed to reencrypt: %v", err)
 	}
 
-	// Decode K
-	kbuff, err := hex.DecodeString(parts[0])
+	outStr, err := encodeReencrypted(xhatenc)
 	if err != nil {
-		return nil, nil, xerrors.Errorf("failed to decode k point: %v", err)
+		return xerrors.Errorf("failed to encode the reencryption data: %v", err)
 	}
 
-	k = suite.Point()
+	fmt.Fprint(ctx.Out, outStr)
+
+	return nil
+}
 
-	err = k.UnmarshalBinary(kbuff)
+func encodeReencrypted(xhatenc kyber.Point) (string, error) {
+	buff, err := xhatenc.MarshalBinary()
 	if err != nil {
-		return nil, nil, xerrors.Errorf("failed to unmarshal k point: %v", err)
+		return "", xerrors.Errorf("failed to marshal xhatenc: %v", err)
 	}
 
-	// Decode C
-	cbuff, err := hex.DecodeString(parts[1])
+	encoded := hex.EncodeToString(buff)
+
+	return encoded, nil
+}
+
+func decodePublicKey(str string) (pk kyber.Point, err error) {
+	pkbuff, err := hex.DecodeString(str)
 	if err != nil {
-		return nil, nil, xerrors.Errorf("failed to decode c point: %v", err)
+		return nil, xerrors.Errorf("failed to decode public key: %v", err)
 	}
 
-	c = suite.Point()
+	pk = suite.Point()
 
-	err = c.UnmarshalBinary(cbuff)
+	err = pk.UnmarshalBinary(pkbuff)
 	if err != nil {
-		return nil, nil, xerrors.Errorf("failed to unmarshal c point: %v", err)
+		return nil, xerrors.Errorf("failed to unmarshal pk: %v", err)
+	}
+
+	return pk, nil
+}
+
+func encodeEncrypted(k kyber.Point, cs []kyber.Point) (string, error) {
+	kbuff, err := k.MarshalBinary()
+	if err != nil {
+		return "", xerrors.Errorf("failed to marshal k: %v", err)
+	}
+
+	encoded := hex.EncodeToString(kbuff)
+
+	for _, c := range cs {
+		cbuff, err := c.MarshalBinary()
+		if err != nil {
+			return "", xerrors.Errorf("failed to marshal c: %v", err)
+		}
+		encoded += separator + hex.EncodeToString(cbuff)
+	}
+
+	return encoded, nil
+}
+
+func decodeEncrypted(str string) (k kyber.Point, cs []kyber.Point, err error) {
```

Comment:
the function is quite long, I'd avoid using named return variables, declaring them properly and preallocating the right size for `cs`.